### PR TITLE
Don't change contents of values in Hashes

### LINKED ISSUE #66: Cannot parse variable with `=>`
We've encountered an issue when trying to send an email using a dynamic template, where one o the dynamic template variables contain a `=>`.

## How to reproduce

The following test case demonstrates the issue:

          it "should not mess up contents" do
            custom_args = {"text" => "line with a => in it"}
            mail["custom_args"] = custom_args
            mailer.deliver!(mail)
            expect(client.sent_mail['custom_args']).to eq({"text" => "line with a => in it"})
          end

The test case ends up failing with a 

    Failure/Error: JSON.parse(text, symbolize_names: symbolize)

     JSON::ParserError:
       767: unexpected token at '{"text":"line with "a":in it"}'
     # ./lib/sendgrid_actionmailer.rb:142:in `json_parse'
     # ./lib/sendgrid_actionmailer.rb:165:in `add_send_options'
     # ./lib/sendgrid_actionmailer.rb:33:in `deliver!'
     # ./spec/lib/sendgrid_actionmailer_spec.rb:323:in `block (5 levels) in <module:SendGridActionMailer>'

See https://github.com/eddiezane/sendgrid-actionmailer/pull/65/commits/2df06f5a5cfcc62ca9e9f4d355a5e43161dd017d to for detailed changes.

## Debugging

The immediate issue seems to be that `SendGridActionMailer::DeliveryMethod#json_parse` tries to convert the text representation of a Hash to JSON in order to parse it back to a Hash. In doing so, it changes all occurrences of `"=>"` to `":"`.

The issue seems to have been introduced in https://github.com/eddiezane/sendgrid-actionmailer/commit/413db5685f782547bf7db0f019d1749660d0fc10.

### PR DESCRIPTION
**Pull request by koppen**
Fixes #66 

This changes `SendGridActionMailer::DeliveryMethod#json_parse` so that it no longer converts a `Hash` to a `String`, transforming the `String` into something that resembles JSON, then uses `JSON.parse` to construct `Hash`. 

If `json_parse` is given a `Mail::Field` whose value is really a `Hash`, it will simply return that `Hash` instead of jumping through the above hoops, potentially breaking the contents of the values in the process, causing errors during JSON parsing.

### CODE REVIEW

File: lib/sendgrid_actionmailer.rb [hunk 1]
```diff
@@ -163,7 +163,18 @@ def add_content(sendgrid_mail, mail)
     end
 
     def json_parse(text, symbolize=true)
```
Comment (koppen) [inline lib/sendgrid_actionmailer.rb]: I wonder if this method should be renamed. It seems to be more focused on converting values to a Hash rather than parsing JSON, the actual JSON parsing is really an implementation detail - a detail which isn't used in all cases now.

File: lib/sendgrid_actionmailer.rb [hunk 2]
```diff
@@ -188,12 +199,12 @@ def add_send_options(sendgrid_mail, mail)
          sendgrid_mail.template_id = mail['template_id'].to_s
       end
       if mail['sections']
-        json_parse(mail['sections'].value, false).each do |key, value|
+        json_parse(mail['sections'], false).each do |key, value|
```
Comment (koppen) [inline lib/sendgrid_actionmailer.rb]: We might be able to not call `json_parse` at all here, although we might still need to do some key transformations to ensure the keys are still `Strings`, though.

**Comment by koppen** [comment]
Hm, so it looks like getting this to be backwards compatible with mail gems before 2.7 is going to be tricky...

Basically it seems the root cause of the issue is the fact that Mail::Field on older versions of mail converts all values to String when the field is instantiated, and we have no way of getting the original value - besides trying to manually deserialize that String.

And I assume we pass all values around as Mail::Fields because that’s the hooks we have into ActionMailer so that part is a given.

I’ll have to think this over (feedback and ideas appreciated), perhaps the best approach would be to fix the issue for only mail 2.7 plus for now.

**Comment by tyrauber** [comment]
@koppen, Yeah, this is tricky. Your understanding of the root cause is correct. I remember having to add some complexity due to changes between mail versions in order to maintain backwards compatibility.

But, I am all for finding a better, simpler, more robust approach. (That gsub regex makes me uncomfortable.)  Perhaps a wrapper method that decodes Mail::Field depending upon Mail version?  Open to any and all ideas.

**Comment by koppen** [comment]
So the hardline version is to simply require mail >= 2.7, bump the gem to 3.0 because that breaks backwards compatibility, and get rid of the json_parse method in its entirety (I think that should be possible, but I haven't verified this). This would make for a better gem going forward, but it does leave some people stranded without an upgrade path, potentially.

The softball version would be to wrap the two failing examples in a guard clause checking if we're running on Mail 2.7. That means we're effectively going through different code paths based on the version of Mail, and that users on Mail 2.6 and older will still have the problem. Admittedly, they've had that for quite a while so that probably doesn't change much for them.

**Comment by tyrauber** [comment]
If using unparsed_value in Mail 2.7 allows us to avoid having to do all the hash translation and regex parsing in json_parse, then I think that makes a stronger, more resilient gem. It would require people update to Rails 5, but people on Rails 4 could still use previous gems. We could cut a release, so the documentation is available for the last Rails 4 supported build. We just couldn't support bug fixes to previous versions (although, someone could always fork).

That to me seems like the most sensible path forward.  Especially, if it doesn't break the current gem api / documentation. It would solve #66 and #70.

**Comment by koppen** [comment]
I absolutely agree. This would require a new major release (following semver and all), but I suspect the benefits outweight the downsides.

We have a branch that does pretty much what you suggest at https://github.com/substancelab/sendgrid-actionmailer/tree/eventzonen, which we have been running in production for a couple of weeks by now.

Another nice benefit that we've discovered is the fact that we can now include all emojis in our email templates, which is very important 😁

**Comment by tyrauber** [comment]
@koppen That's great! Is the documentation and test coverage updated? Mind creating a pull release so we can review the diff.

**Comment by koppen** [comment]
Hm, the branch currently doesn't apply cleanly. I should be able to update this PR with the necessary changes. Will get on that later this week.

**Comment by koppen** [comment]
This now applies cleanly and all specs pass on all the tested rubies. However, it _is_ a breaking change as we'll no longer work on mail gems before 2.7.

1. VERSION has prematurely been upped to 3.0.
2. Test for mail 2.5 and mail 2.6 have been removed from CI.
3. CHANGELOG has been updated to reflect this.

**Comment by tyrauber** [comment]
Hi @koppen, thank you for doing this. I have a few questions:

- Do we still need the parse_json method? Can't we just use unparsed_value directly?
- If we are mandating Mail 2.7, we don't need to test for unparsed_value anymore, correct?
- Can't we eliminate the [text based regex](https://github.com/eddiezane/sendgrid-actionmailer/pull/67/files#diff-5f8621c144b574135a554c09fcdc5a0bR186)?
- Can't we eliminate the text based [JSON.parse](https://github.com/eddiezane/sendgrid-actionmailer/pull/67/files#diff-5f8621c144b574135a554c09fcdc5a0bR187)?
- Do we need [transform_keys](https://github.com/eddiezane/sendgrid-actionmailer/pull/67/files#diff-5f8621c144b574135a554c09fcdc5a0bR305-R321), or should we just mandate Ruby 2.5?

Basically, I am wondering if we can remove or rename parse_json and [simplify this whole block](https://github.com/eddiezane/sendgrid-actionmailer/pull/67/files#diff-5f8621c144b574135a554c09fcdc5a0bR176-R187)?

**Comment by koppen** [comment]
> Do we need [transform_keys](https://github.com/eddiezane/sendgrid-actionmailer/pull/67/files#diff-5f8621c144b574135a554c09fcdc5a0bR305-R321), or should we just mandate Ruby 2.5?

That's actually a good point. I went with trying to break as little as possible, but I guess if we're moving towards a major release anyway, it might be a decent opportunity to update Ruby requirements as well.

> If we are mandating Mail 2.7, we don't need to test for unparsed_value anymore, correct?

In theory, no, but we still pass Strings and other non-`Mail::Field` values to the method, at least in specs. We need to test for those, I suppose.

> Do we still need the parse_json method? Can't we just use unparsed_value directly?

I tried removing it, but other examples started failing, so I figured it might be used/needed for something still. It's definitely worth revisiting, though, but perhaps it's a separate change, so this PR could be kept more focused? Not that it matters much either way, but I figured you'd want as few changes as possible at once 😀

**Comment by tyrauber** [comment]
hi @koppen, I extended #67 and removed the json_parse method and janky regex in favor of unparsed_value. I also updated the transform_hash method to handle nested hashes.  Due to the json translation built into sendgrid-ruby, hash translation was still required - primarily for testing and accommodating nested hashes.  Backwards compatible with previous versions of Ruby. Readme stays the same. Specs are green.  Would love your review.
